src/librustc_typeck/check/wfcheck.rs

```diff
@@ -368,14 +370,22 @@ fn check_where_clauses<'a, 'gcx, 'fcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
     let mut substituted_predicates = Vec::new();
 
     let generics = tcx.generics_of(def_id);
-    let is_our_default = |def: &ty::TypeParameterDef|
-                            def.has_default && def.index >= generics.parent_count() as u32;
+    let is_our_default = |def: &ty::GenericParamDef| {
+        def.to_type().has_default && def.index >= generics.parent_count as u32
+    };
 
     // Check that concrete defaults are well-formed. See test `type-check-defaults.rs`.
     // For example this forbids the declaration:
     // struct Foo<T = Vec<[u32]>> { .. }
     // Here the default `Vec<[u32]>` is not WF because `[u32]: Sized` does not hold.
-    for d in generics.types.iter().cloned().filter(is_our_default).map(|p| p.def_id) {
+    for d in generics.params.iter().filter_map(|param| {
+        if let GenericParamDefKind::Type(_) = param.kind {
+            if is_our_default(&param) {
+                return Some(param.def_id);
+            }
+        }
+        None
```

Comment:
I prefer the `filter` + `map` approach (`is_our_default` could handle the `GenericParamDefKind::Type(_)` check) *or* this could go in at the start of the `for`.